packages/utils/src/dal/mikro-orm/mikro-orm-repository.ts

```diff
@@ -420,6 +435,329 @@ export function mikroOrmBaseRepositoryFactory<T extends object = object>(
+        throw new Error(
+          `Nonexistent relations were passed during upsert: ${nonexistentRelations}`
+        )
+      }
+
+      // We want to response with all the data including the IDs in the same order as the input. We also include data that was passed but not processed.
+      const reconstructedResponse: any[] = []
+      const originalDataMap = new Map<string, T>()
+      const upsertsPerType: Record<string, any[]> = {}
+
+      // Create only the top-level entity without the relations first
+      const toUpsert = normalizedData.map((entry) => {
+        // Make a copy of the data and remove undefined fields. The data is already a POJO due to the serialization above
+        const entryCopy = JSON.parse(JSON.stringify(entry))
+        const reconstructedEntry: any = {}
+
+        allRelations?.forEach((relation) => {
+          reconstructedEntry[relation.name] = this.handleRelationAssignment_(
+            manager,
+            upsertConfig,
+            allRelationsToHandle,
+            relation,
+            entryCopy,
+            upsertsPerType
+          )
+        })
+
+        const mainEntity = this.getEntityWithId(manager, entity.name, entryCopy)
+        reconstructedResponse.push({ ...mainEntity, ...reconstructedEntry })
+        originalDataMap.set(mainEntity.id, entry)
+
+        return mainEntity
+      })
+
+      // The order of insert is: many-to-one, main entity, one-to-many, many-to-many
+      await this.upsertManyToOneRelations_(manager, upsertsPerType)
+      const upsertedTopLevelEntities = await this.upsertMany_(
+        manager,
+        entity.name,
+        toUpsert
+      )
+
+      await promiseAll(
+        upsertedTopLevelEntities.map(async (entityEntry, i) => {
+          const originalEntry = originalDataMap.get((entityEntry as any).id)!
+          const reconstructedEntry = reconstructedResponse[i]
+
+          await promiseAll(
+            allRelations?.map(async (relation) => {
+              const relationName = relation.name as keyof T
+              if (!upsertConfig.relationsToUpsert?.includes(relationName)) {
+                return
+              }
+
+              // TODO: Handle ONE_TO_ONE
+              // One to one and Many to one are handled outside of the assignment as they need to happen before the main entity is created
+              if (
+                relation.reference === ReferenceType.ONE_TO_ONE ||
+                relation.reference === ReferenceType.MANY_TO_ONE
+              ) {
+                return
+              }
+
+              reconstructedEntry[relationName] =
+                await this.assignCollectionRelation_(
+                  manager,
+                  { ...originalEntry, id: (entityEntry as any).id },
+                  relation
+                )
+              return
+            })
+          )
+        })
+      )
+
+      // // We want to populate the identity map with the data that was written to the DB, and return an entity object
```

Comment:
Leaving this for a follow-up PR as it isn't as simple as uncommenting it and it isn't required per-se.